print-geonet.c

```diff
@@ -277,24 +270,24 @@ static void gn_common_header_decode_from_bytes(netdissect_options *ndo, const u_
 	uint64_t value = GET_BE_U_8(*bp);
 	*bp += 8;
 	*length -= 8;
-	*next_header = (value >> (4 + 7 * 8)) & 0x3;
-	reserved = (value >> 7 * 8) & 0xF;
-	*header_type = (value >> (4 + 6 * 8)) & 0xF;
-	*header_subtype = (value >> 6 * 8) & 0xF;
-
-	uint8_t tc_encoded = (value >> 5 * 8) & 0xFF;
-	tc_scf = (tc_encoded >> 7) & 0x01;
-	tc_channel_offload = (tc_encoded >> 6) & 0x01;
-	tc_id = tc_encoded & 0x3F;
-
-	flags = (value >> 4 * 8) & 0xFF;
-	pl = (value >> 2 * 8) & 0xFFFF;
-	mhl = (value >> 8) & 0xFF;
-	reserved2 = value & 0xFF;
-
-	const char *next_header_text = common_header_next_header_text_from_bytes(*next_header);
-	const char *header_type_text = header_type_text_from_bytes(*header_type, *header_subtype);
-	const char *flags_text = flags_text_from_bytes(flags);
+	*next_header = (value >> (4 + SEVEN_BYTES)) & THREE_BITS_MASK;
+	reserved = (value >> SEVEN_BYTES) & FOUR_BITS_MASK;
+	*header_type = (value >> (4 + SIX_BYTES)) & FOUR_BITS_MASK;
+	*header_subtype = (value >> SIX_BYTES) & FOUR_BITS_MASK;
+
+	uint8_t tc_encoded = (value >> FIVE_BYTES) & EIGHT_BITS_MASK;
+	tc_scf = (tc_encoded >> 7) & ONE_BIT_MASK;
+	tc_channel_offload = (tc_encoded >> 6) & ONE_BIT_MASK;
+	tc_id = tc_encoded & SIX_BITS_MASK;
+
+	flags = (value >> FOUR_BYTES) & EIGHT_BITS_MASK;
+	pl = (value >> TWO_BYTES) & SIXTEEN_BITS_MASK;
+	mhl = (value >> ONE_BYTE) & EIGHT_BITS_MASK;
+	reserved2 = value & EIGHT_BITS_MASK;
+
+	const char *next_header_text = tok2str(common_header_next_header_values, "Unknown", *next_header);
+	const char *header_type_text = tok2str(header_type_tok, "Unknown", HT_HST(*header_type, *header_subtype));
+	const char *flags_text = tok2str(flags_text_from_bytes, "Unknown", flags);
```

Comment:
Likewise, the Common Header should have been a structure or a series of 1-octet and 2-octet fetches, not a 64-bit fetch. Section 9.7 Ibid.